pf/tfbridge/ids.go

```diff
@@ -15,81 +15,28 @@
-// present. This code encapsulates extracting ID and all the error handling. It tries to fail early at the static level
-// if a schema for a resource does not specify the expected ID field, but may also fail later at runtime if the data
-// does not have an ID or it is of the wrong type.
-func newIDExtractor(ctx context.Context, typeName string, schema pfutils.Schema) (idExtractor, error) {
-	idPath := path.Root("id")
-
-	idAttr, diags := schema.AttributeAtPath(ctx, idPath)
-	if diags.HasError() {
-		msg := "Cannot bridge Terraform resource %q to Pulumi: " +
-			"schema does not define the required %q attribute"
-		return idExtractor{}, fmt.Errorf(msg, typeName, "id")
-	}
-
-	idAttrType := idAttr.GetType().TerraformType(ctx)
-	if !idAttrType.Is(tftypes.String) {
-		msg := "Cannot bridge Terraform resource %q to Pulumi: " +
-			"the %q attribute has type %s but only %s is supported"
-		return idExtractor{}, fmt.Errorf(msg, typeName, "id", idAttrType.String(), tftypes.String.String())
-	}
-
-	return idExtractor{typeName: typeName}, nil
-}
-
-func (ie idExtractor) extractID(state tftypes.Value) (string, error) {
-	typeName := ie.typeName
-	idAttrPath := tftypes.NewAttributePath().WithAttributeName("id")
-	idValue, gotIDValue, err := valueAtPath(idAttrPath, state)
-	if err != nil {
-		return "", fmt.Errorf(
-			"Cannot extract ID from %q resource state: %w", typeName, err)
+func extractID(
+	resname string, info *tfbridge.ResourceInfo, state resource.PropertyMap,
+) (resource.ID, error) {
+	if info != nil && info.ComputeID != nil {
+		return info.ComputeID(state)
 	}
-	if !gotIDValue {
-		return "", fmt.Errorf(
-			"Cannot extract ID from %q resource state: ID attribute is missing", typeName)
-	}
-	var idString string
-	if err := idValue.As(&idString); err != nil {
-		return "", fmt.Errorf(
-			"Cannot extract ID from %q resource state, expecting a string: %w", typeName, err)
-	}
-	if idString == "" {
-		return "", fmt.Errorf(
-			"Cannot extract ID from %q resource state: ID cannot be empty", typeName)
-	}
-	return idString, nil
-}
-
-// Drills down into a Value. Returns the found value and a flag indicating if it was found or not.
-func valueAtPath(p *tftypes.AttributePath, root tftypes.Value) (tftypes.Value, bool, error) {
-	result, _, err := tftypes.WalkAttributePath(root, p)
-	if err == tftypes.ErrInvalidStep {
-		return tftypes.Value{}, false, nil // not found
-	}
-	if err != nil {
-		return tftypes.Value{}, false, err // error
-	}
-	resultValue, ok := result.(tftypes.Value)
+	idValue, ok := state["id"]
+	c := fmt.Sprintf(". If special identity handling is needed, consider customizing "+
+		"ResourceInfo.ComputeID for the %s resource", resname)
 	if !ok {
-		return tftypes.Value{}, false, fmt.Errorf(
-			"Expected a value of type tftypes.Value but got: %v", result)
+		return "", fmt.Errorf("Resource state did not contain an 'id' property" + c)
+	}
+	if !idValue.IsString() {
```

Comment:
If `idValue` is computed, an output, or a secret then this will break. If we are sure that this will not happen, can we `contract.Assertf` that and explain why?

Reply:
I was habitually using the error channel but we can try asserts, especially if the tfgen-time checks should have eliminated some possibilities here. Panics are less user friendly since the provider doesn't recover other resource operations that may be in flight, but they certainly get our attention through reporting.

Reply:
My concern here was documentation. I read `if v.IsComputed() { return "", fmt.Errorf("v is computed") }`, I assume that `v` could be computed, but should not be. If I read `contract.Assertf(!v.IsComputed(), "v is computed")` then I assume that `v` is know to not be computed, and we are calling this out.

If some subset of valid TF providers can return a value that doesn't pass this check, we shouldn't assert.

Reply:
Well idValue.IsString() check for that but sure I can add another assert. The state should not have computers or unknowns. It's post-create state or post-refresh state it should only be dealing with known values. This being Go and not Haskell we don't have a type-level way to express this.